File: 1941E-rudolfAndkBridges.py
# Author : AlifSrSE
# Date : 2025-03-11
# Problem link : https://codeforces.com/contest/1941/problem/E

import sys

input = sys.stdin.readline
INF = 0x3f3f3f3f3f3f3f3f

class SegmentTree:
    def __init__(self, n):
        self.n = n
        self.tree = [INF] * (4 * n)
    
    def update(self, k, l, r, x, v):
        if l == r and l == x:
            self.tree[k] = v
            return
        if x < l or x > r:
            return
        mid = (l + r) // 2
        self.update(2 * k, l, mid, x, v)
        self.update(2 * k + 1, mid + 1, r, x, v)
        self.tree[k] = min(self.tree[2 * k], self.tree[2 * k + 1])
    
    def query(self, k, l, r, x, y):
        if y < l or x > r:
            return INF
        if x <= l and r <= y:
            return self.tree[k]
        mid = (l + r) // 2
        return min(self.query(2 * k, l, mid, x, y),
                  self.query(2 * k + 1, mid + 1, r, x, y))

def main():
    T = int(input())
    for _ in range(T):
        n, m, num, d = map(int, input().split())
        ans = [0] * (n + 1)
        f = [0] * (m + 1)  
        seg = SegmentTree(m)
        
        for i in range(1, n + 1):
            a = [0] + list(map(int, input().split()))
            f[1] = 1
            seg.tree = [INF] * (4 * m)
            seg.update(1, 1, m, 1, 1)
            
            for j in range(2, m + 1):
                f[j] = seg.query(1, 1, m, max(j - d - 1, 1), max(j - 1, 1)) + a[j] + 1
                seg.update(1, 1, m, j, f[j])
            
            ans[i] = f[m]
        
        for i in range(1, n + 1):
            ans[i] += ans[i - 1]
        
        min_sum = INF
        for i in range(1, n - num + 2):
            min_sum = min(min_sum, ans[i + num - 1] - ans[i - 1])
        
        print(min_sum)

if __name__ == "__main__":
    main()

File: 1941E-rudolfAndKBridges.java
// Author : AlifSrSE
// Date : 2025-03-12
// Problem link : https://codeforces.com/contest/1941/problem/E


import java.io.*;
import java.util.*;

public class Main {
    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        public FastScanner(InputStream stream) {
            br = new BufferedReader(new InputStreamReader(stream));
        }

        String next() {
            while (st == null || !st.hasMoreTokens()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }
    }

    static long makeBridge(List<Long> a, int d) {
        int len = a.size();
        if (len <= d) return a.get(0) + a.get(len - 1) + 2;

        long[] dp = new long[len];
        dp[0] = a.get(0) + 1;

        Deque<long[]> dq = new ArrayDeque<>();
        dq.add(new long[]{dp[0], 0});

        for (int i = 1; i < len; i++) {
            while (!dq.isEmpty() && dq.peekFirst()[1] < i - d - 1)
                dq.pollFirst();

            dp[i] = dq.peekFirst()[0] + a.get(i) + 1;
            while (!dq.isEmpty() && dq.peekLast()[0] >= dp[i])
                dq.pollLast();
            dq.add(new long[]{dp[i], i});
        }

        return dp[len - 1];
    }

    static void solve(int t, int total, FastScanner sc, PrintWriter out) {
        int n = sc.nextInt();
        int m = sc.nextInt();
        int k = sc.nextInt();
        int d = sc.nextInt();

        List<List<Long>> f = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            List<Long> row = new ArrayList<>(m);
            for (int j = 0; j < m; j++)
                row.add(sc.nextLong());
            f.add(row);
        }

        long[] b = new long[n];
        for (int i = 0; i < n; i++)
            b[i] = makeBridge(f.get(i), d);

        long sum = 0;
        long minCost = Long.MAX_VALUE;
        for (int i = 0; i < k; i++)
            sum += b[i];
        minCost = sum;

        for (int i = k; i < n; i++) {
            sum = sum - b[i - k] + b[i];
            minCost = Math.min(minCost, sum);
        }

        out.print(minCost);
        if (t != total) out.println();
    }

    public static void main(String[] args) throws IOException {
        FastScanner sc = new FastScanner(System.in);
        PrintWriter out = new PrintWriter(System.out);
        
        int t = sc.nextInt();
        for (int i = 1; i <= t; i++)
            solve(i, t, sc, out);
        
        out.flush();
    }
}